common/fixation_entry.go

```diff
@@ -10,148 +10,137 @@ import (
+//
+// 4. VersionedStore keeps a reference count of versioned of entries, and when the
+// count reaches 0 it marks them for deletion. The actual deletion takes place after
+// a fixed number of epochs has passed.
 
 type VersionedStore struct {
 	storeKey sdk.StoreKey
 	cdc      codec.BinaryCodec
+	prefix   string
 }
 
 // AppendEntry adds a new entry to the store
-func (vs VersionedStore) AppendEntry(ctx sdk.Context, fixationKey string, index string, block uint64, entryData codec.ProtoMarshaler) error {
-	// delete old entries
-	vs.deleteStaleEntries(ctx, fixationKey)
-
+func (vs VersionedStore) AppendEntry(ctx sdk.Context, index string, block uint64, entryData codec.ProtoMarshaler) error {
 	// get the latest entry for this index
-	latestEntry := vs.getUnmarshaledEntryForBlock(ctx, fixationKey, index, block, types.DO_NOTHING)
+	latestEntry := vs.getUnmarshaledEntryForBlock(ctx, index, block, types.DO_NOTHING)
 
 	// if latest entry is not found, this is a first version entry
-	firstVersion := false
 	if latestEntry == nil {
-		firstVersion = true
-	}
-
-	// make sure the new entry's block is not smaller than the latest entry's block
-	if !firstVersion && block < latestEntry.GetBlock() {
-		return utils.LavaError(ctx, ctx.Logger(), "AppendEntry_block_too_early", map[string]string{"latestEntryBlock": strconv.FormatUint(latestEntry.GetBlock(), 10), "block": strconv.FormatUint(block, 10), "index": index, "fixationKey": fixationKey}, "can't append entry, earlier than the latest entry")
-	}
+		vs.SetEntryIndex(ctx, index)
+	} else {
+		// make sure the new entry's block is not smaller than the latest entry's block
+		if block < latestEntry.GetBlock() {
+			return utils.LavaError(ctx, ctx.Logger(), "AppendEntry_block_too_early", map[string]string{"latestEntryBlock": strconv.FormatUint(latestEntry.GetBlock(), 10), "block": strconv.FormatUint(block, 10), "index": index, "vs.prefix": vs.prefix}, "can't append entry, earlier than the latest entry")
+		}
 
-	// if the new entry's block is equal to the latest entry, overwrite the latest entry
-	if !firstVersion && block == latestEntry.GetBlock() {
-		return vs.SetEntry(ctx, fixationKey, index, block, entryData)
+		// if the new entry's block is equal to the latest entry, overwrite the latest entry
+		if block == latestEntry.GetBlock() {
+			return vs.SetEntry(ctx, index, block, entryData)
+		}
 	}
 
 	// marshal the new entry's data
 	b := vs.cdc.MustMarshal(entryData)
 	// create a new entry and marshal it
-	entry := types.Entry{Index: index, Block: block, Data: b, References: 0}
+	entry := types.Entry{Index: index, Block: block, Data: b, Refcount: 0}
 	bz := vs.cdc.MustMarshal(&entry)
 
 	// get the relevant store
-	store := prefix.NewStore(ctx.KVStore(vs.storeKey), types.KeyPrefix(types.EntryKey+fixationKey+index))
-	byteKey := types.KeyPrefix(createEntryKey(index, fixationKey, block))
+	store := prefix.NewStore(ctx.KVStore(vs.storeKey), types.KeyPrefix(types.EntryKey+vs.prefix+index))
```

Comment:
types.EntryKey+vs.prefix+index move this to a function, we use it all over the place

Reply:
done